oeps/oep-0007.rst

```diff
@@ -0,0 +1,523 @@
+
+.. code:: python3
+
+    if six.PY2:  # or future.PY2
+        do_python2_thing()
+    else:
+        do_python3_thing()
+
+Do not explicitly call out ``six.PY3`` or ``future.PY3``.  This should be more
+future-compatible with a potential future Python 4.
+
+
+Changing Coding Practice
+++++++++++++++++++++++++
+
+Changing internal code practices to ease conversion will require a
+three-pronged approach of documentation, socialization, and tooling.  To start,
+we need to update the official edX code style guide to mandate compatible code
+practices.  To socialize these practices among our engineers, we will announce
+our efforts to migrate to Python 3 during an engineering all-hands meeting,
+offer a workshop in writing compatible code, and promote awareness of
+incompatibilities during code reviews.  Additionally, we will host regular
+Python 3 office hours to help answer questions and troubleshoot problems that
+arise during migration.
+
+Appropriate tooling will help.  Tests should be configured to run under both
+Python 2 or 3 using tox.  As it would double the cost of running
+tests to run both all the time, we may only want to enable Python 3 tests in
+the CI environment for repos that are actively being migrated.  We still want
+to ensure that Python 3 compatibility is improving on other projects, so a
+lightweight metric that can be run every time would also be useful, such as
+running pylint with the ``caniusepython3.pylint_checker`` extension.  Making these checks
+mandatory in a similar way to our current quality will ensure that
+compatibility is improving (TODO: timeline for this?).
+
+
+Migrating projects
+++++++++++++++++++
+
+We should be able to migrate individual applications to Python 3 independently.
+To begin with, we should pilot the process using a relatively small (but
+complex enough to provide useful information) IDA.  As we go, we will document
+the process, find pain points, figure out ways of dealing with them, and
+continue to improve our process.
+
+For a given project, steps are:
+
+1. Turn on caniusepy3k linting, and reset the lint error cap.
+2. Turn on tox testing in Python 3, but allow the tests to fail.
+3. Reduce the number of lint errors to zero, lowering lint error cap as you go.
+   Optionally, use ``futurize``, phase 1 to automate the first stage of the
+   conversion process.
+4. Reduce the number of failing tests to zero.  This may involve updating
+   dependent libraries to Python 3 compatibile versions.  It will almost
+   certainly involve normalizing text handling.
+5. Making failing Python 3 tests fail the build.
+6. Deploy the project in Python 3.
+7. Stop testing under Python 2.
+
+
+Order of migrations
++++++++++++++++++++
+
+* IDAs that we want to continue supporting in the future
+  + Old IDAs (that we want to replace) should not be upgraded, but we will need
+    to prioritize replacement to occur during the migration timeframe.
```

Comment:
The nested bullets here and in the third bullet point need to be [separated by blank lines](http://www.sphinx-doc.org/en/1.5.1/rest.html#lists-and-quote-like-blocks) from the parent list items to be formatted correctly